Fix/warn if waiting too long to connect to server

schoinh opened the pull request:
Problem
=======
Closes #169 

Solution
========
The timeout for connecting to the server was already built into `RemoteSimulator` but the error message was not getting propagated up as a `FrontEndError`. So I made that error bubble up from `connectToRemoteServer` to `startRemoteTrajectoryPlayback` where it would be thrown as a `FrontEndError` that would be displayed in the UI. I also added some tests.

While working on the above I refactored `connectToRemoteServer` to be more testable and hopefully a little clearer in general.

Thanks to @meganrm for help writing tests!

## Type of change
Please delete options that are not relevant.

* Bug fix (non-breaking change which fixes an issue)
* This change requires updated or new tests

Steps to Verify:
----------------
1. Test should pass
2. `npm start` and check that networked models load in the viewer testbed with no new problems

Keyfiles (delete if not relevant):
-----------------------
1. src/simularium/RemoteSimulator.ts
2. src/test/RemoteSimulator.test.ts

Thanks for contributing!

schoinh commented on src/constants.ts:
```diff
@@ -19,3 +19,8 @@ export const DEFAULT_CAMERA_SPEC: CameraSpec = {
     fovDegrees: 75,
 };
+
+export const TEST_CONNECTION_SETTINGS = {
```
I feel iffy about this variable name, better ideas welcome

toloudis replied:
I'll propose STAGING_SERVER_SETTINGS ?

schoinh commented on src/simularium/RemoteSimulator.ts:
```diff
@@ -255,53 +260,52 @@ export class RemoteSimulator implements ISimulator {
         return `wss://${this.serverIp}:${this.serverPort}/`;
     }
 
-    private connectToUriAsync(address): Promise<string> {
-        const connectPromise = new Promise<string>((resolve) => {
-            this.connectToUri(address);
-            resolve("Successfully connected to uri!");
-        });
-
-        return connectPromise;
-    }
```
After discussing WebSocket object creation with @EricJIsaac , I thought this function was unnecessary, since it's an async wrapper for a function (`connectToUri`) which doesn't have any async operations?

schoinh commented on src/simularium/RemoteSimulator.ts:
```diff
@@ -255,53 +260,52 @@ export class RemoteSimulator implements ISimulator {
-    public connectToRemoteServer(address: string): Promise<string> {
+    public connectToRemoteServer(
+        address: string,
+        timeout = 1000
```
New parameter `timeout` allows testing with a very short timeout (like 1 instead of 1000 ms), as a way of simulating a slow connection

schoinh commented on src/simularium/RemoteSimulator.ts:
```diff
@@ -255,53 +260,52 @@ export class RemoteSimulator implements ISimulator {
-                const isConnected = await waitForIsConnected();
-                secondsWaited++;
+            /*
+            Initially wait for a max wait time of MAX_WAIT_TIME, then retry connecting 
+            <MAX_RETRIES> time(s). In a retry, only wait for the amount of time 
+            specified as timeout. (Can reset timeWaited to 0 upon retry if we want to
+            wait for MAX_WAIT_TIME in a retry too)
```
Not positive if this is what we want, but I didn't change it

toloudis replied:
It looks to me like we are waiting for a total of MAX_WAIT_TIME over many possible connection retries.  I am led to think that because of line 293 timeWaited+=timeout.  If I am correct, the comment should say something like "wait timeout time, try to reconnect, until a total of MAX_WAIT_TIME has elapsed before giving up"

schoinh replied:
That's what I thought/expected too at first, but if you look at line 296-297 it won't let you get to a connection retry (in line 299) until `timeWaited` is at least at `MAX_WAIT_TIME`, which is what (along with some console logs) led to my interpretation. But this function really needs a test so that we know for sure how long we're making the user wait and how many retries we're allowing. I tried to break it out into a class method so I can test it but it turned into a tangled mess, so I reverted it 😅 But seeing your comment makes me want to try again, so I'll give it another go at testing this recursion.

toloudis replied:
I see what you mean.  Thanks.  It's basically just polling on a shorter interval (`timeout`) but allowing for a long interval (MAX_WAIT_TIME) for the total time to connect.

meganrm replied:
Yeah, I think the idea overall was that if it had failed to connect after waiting a while, waiting a second time wouldn't help much, the problem isn't a slow connection, but no connection. 

Have you tried to factor out `checkConnection` and pass in the two counters?

schoinh replied:
That makes sense @meganrm . I factored out `checkConnection` but instead of passing in the counters as params I turned them into class properties. Let me know if there's a good reason to make them function params instead.

schoinh commented on src/simularium/RemoteSimulator.ts:
```diff
@@ -314,7 +318,7 @@ export class RemoteSimulator implements ISimulator {
     }
 
-    private sendWebSocketRequest(jsonData, requestDescription): void {
+    public sendWebSocketRequest(jsonData, requestDescription: string): void {
```
Needed this public for testing

toloudis replied:
I wonder if there's a way to let a thing stay private but make it public for unit tests

schoinh replied:
I couldn't find a way to do that, but I modified the test so that we don't have to access this method.

toloudis replied:
Cool.  You can do it in C++ by using the "friend" keyword which allows other classes access to my class's private members.

schoinh replied:
Whaaaaat that's pretty cool.

schoinh commented on src/simularium/RemoteSimulator.ts:
```diff
@@ -404,12 +408,16 @@ export class RemoteSimulator implements ISimulator {
+                    "Start Trajectory File Playback"
+                );
+            })
+            .catch((error) => {
+                throw new FrontEndError(error);
+            });
```
This is the real fix that should get the error message "working"

schoinh commented on src/test/RemoteSimulator.test.ts:
```diff
@@ -0,0 +1,73 @@
+            );
+            expect(message).toEqual(CONNECTION_SUCCESS_MSG);
+        });
+        test("emits error if connecting to server takes longer than allotted time", async () => {
+            const simulator = new RemoteSimulator(TEST_CONNECTION_SETTINGS);
+            const veryShortTimeout = 1;
+
+            try {
+                await simulator.connectToRemoteServer(
+                    simulator.getIp(),
+                    veryShortTimeout
+                );
+            } catch (error) {
+                expect(error).toEqual(new Error(CONNECTION_FAIL_MSG));
+            }
+        });
```
Key test number 1

schoinh commented on src/test/RemoteSimulator.test.ts:
```diff
@@ -0,0 +1,73 @@
+            );
+            expect(request).toBeCalled();
+        });
+        test("throws error emitted by connectToRemoteServer as a FrontEndError if connection fails", async () => {
+            const simulator = new RemoteSimulator(TEST_CONNECTION_SETTINGS);
+            jest.spyOn(simulator, "connectToRemoteServer").mockRejectedValue(
+                new Error("Mock error message")
+            );
+
+            try {
+                await simulator.startRemoteTrajectoryPlayback(
+                    "endocytosis.simularium"
+                );
+            } catch (error) {
+                expect(error).toEqual(new FrontEndError("Mock error message"));
+            }
+        });
```
Key test number 2

toloudis commented on src/test/RemoteSimulator.test.ts:
```diff
@@ -0,0 +1,73 @@
+            const simulator = new RemoteSimulator({});
+            expect(simulator.socketIsValid()).toBe(false);
+
+            simulator.connectToUri(simulator.getIp());
```
is this testing a real live internet connection?

schoinh replied:
Yeah, it's actually connecting to our staging server. Is that ok?

toloudis replied:
It just means you're no longer testing just simularium-viewer, you're now testing that the staging server and the internet connectivity to it are both alive and well.  There are so many reasons why it could fail.  I would resist doing this in the unit tests here.  Better if you can mock it somehow.  But if others prefer to have this be a true integration test, we can think about it that way.

meganrm replied:
I think it's already mocked in the DummyRemoteSimulator ?

schoinh replied:
Yeah that's a good point. I'll look into using the DummyRemoteSimulator or using Jest to mock some return values like I did in what I called the "key test number 2". I think I would prefer the second option because wouldn't it be better to test the RemoteSimulator class (even if some of its methods are mocked) rather than the DummyRemoteSimulator?

toloudis replied:
I tend to agree, what you really want to mock is the back end, and not the RemoteSimulator itself.  Which i suppose is what the DummyRemoteSimulator is supposed to be trying to do, but it's extending RemoteSimulator.

toloudis commented:
Looking at this code, it strikes me that a good maintenance task on this repo would be to split apart the "viewing" code from the "streaming/downloading" code into separate subdirectories and have them talk to each other through a well thought out interface.  Something not to change any functionality but to start to decouple the two concerns a bit more.  Ideally this would organize the code better but have no functional or performance impact.

schoinh commented:
I need to do a bit more cleanup

toloudis commented on src/simularium/ISimulator.ts:
```diff
@@ -18,7 +18,7 @@ export interface ISimulator {
      * Connect
      * */
-    connectToUri(uri: string): void;
+    createWebSocket(uri: string): void;
```
I hope this can be removed from the ISimulator interface and it's really just private to RemoteSimulator.

schoinh replied:
Removed from ISimulator, ClientSimulator, and LocalSimulator and seems like everything still works fine.

schoinh commented:
Changes since @meganrm and @toloudis 's review: https://github.com/allen-cell-animated/simularium-viewer/pull/173/files/52903c395dae92977e6057e67feede30e32aa322..db1f0c55a73ffa9ec2e39f6ec6b5df2b0b71fbb7

- Factored out `checkConnection` as a class method that's more easily testable
- Used mock return values in all tests instead of real ones so that the tests aren't affected by any network/server issues
- I reverted the creation of the `TEST_CONNECTION_SETTINGS` constant because it was no longer needed in the tests.
- Renamed `connectToUri` method as `createWebSocket` because I think this is less confusing / more accurate
- I had to add a babel plugin `@babel/plugin-transform-runtime`. Something about me turning `waitForWebSocket` into an async class method was giving me a "regeneratorRuntime is not defined" error, so this was necessary to fix it. More info: https://babeljs.io/docs/en/babel-plugin-transform-runtime#regenerator

toloudis commented on src/simularium/RemoteSimulator.ts:
```diff
@@ -255,56 +266,61 @@ export class RemoteSimulator implements ISimulator {
-            let connectionTries = 1;
-            return startPromise.then(handleReturn);
-        });
+        this.createWebSocket(address);
```
Is it possible to move this line inside of this.checkConnection? (is checkConnection called by anything else?)
